src/chainstate/stacks/index/storage.rs

```diff
@@ -919,154 +646,133 @@ impl TrieFileStorage {
             None => (1024) // don't try to guess _byte_ allocation here.
         };
 
-        let identifier = self.block_map.add_block(bhh);
-        if self.chain_tips.contains(&self.cur_block) {
-            self.chain_tips.remove(&self.cur_block);
-        }
+        trie_sql::remove_chain_tip_if_present(&self.db, &self.cur_block)?;
+
         // this *requires* that bhh hasn't been the parent of any prior
         //   extended blocks.
         // this is currently enforced if you use the "public" interfaces
         //   to marfs, but could definitely be violated via raw updates
         //   to trie structures.
-        self.chain_tips.insert(bhh.clone());
-
-        let trie_buf = TrieRAM::new(bhh, size_hint, identifier, &self.cur_block);
-
-        // create an empty file for this block, so we can't extend to it again
-        let block_dir = TrieFileStorage::block_dir(&self.dir_path, bhh);
-        let block_path = TrieFileStorage::block_path(&self.dir_path, bhh);
-        let block_path_tmp = TrieFileStorage::block_path_tmp(&self.dir_path, bhh);
-        // check if block_path exists
-        match fs::metadata(&block_path) {
-            Ok(_) => {
-                trace!("Block path exists: {:?}", &block_path);
-                return Err(Error::ExistsError);
-            },
-            Err(e) => {
-                if e.kind() != io::ErrorKind::NotFound {
-                    return Err(Error::IOError(e));
-                }
-            }
-        };
-        // check if block_path_tmp exists
-        match fs::metadata(&block_path_tmp) {
-            Ok(_) => {
-                error!("Tried to create index block {:?} twice", bhh);
-                return Err(Error::ExistsError);
-            },
-            Err(e) => {
-                if e.kind() != io::ErrorKind::NotFound {
-                    return Err(Error::IOError(e));
-                }
-            }
-        };
-
-        fs::create_dir_all(block_dir)
-            .map_err(Error::IOError)?;
+        trie_sql::add_chain_tip(&self.db, bhh)?;
```

Comment:
I think the above requirement is enforced already by the `marf_data` table's `UNIQUE` constraint on `block_hash`?

Reply:
Re: this add_chain_tip call only being valid if there hasn't already been a block that has this as it's parent: I think you could manually add such a block without violating the unique constraint.

Regarding tracking chain_tips at all -- I'll take a look at what removing the chain_tips access function does, but I suspect you're right -- I think chain_tips was used before the chainstate::db module was fleshed out, and can likely be removed.

Reply:
Removed in 7116618